pkg/cmd/cli/secrets/source.go

```diff
@@ -0,0 +1,213 @@
+				cmdutil.CheckErr(cmdutil.UsageError(c, err.Error()))
+			}
+
+			if err := o.Validate(); err != nil {
+				cmdutil.CheckErr(cmdutil.UsageError(c, err.Error()))
+			}
+
+			if err := o.CreateSourceSecret(); err != nil {
+				cmdutil.CheckErr(err)
+			}
+		},
+	}
+
+	cmd.Flags().StringVar(&o.Username, "username", "", "username for Git authentication")
+	cmd.Flags().StringVar(&o.Password, "password", "", "password for Git authentication")
+	cmd.Flags().StringVar(&o.Token, "token", "", "token for Git authentication")
+	cmd.Flags().StringVar(&o.CertificatePath, "cert-path", "", "path to the certificate file")
+	// cmd.Flags().StringVar(&o.GitConfigPath, "gitconfig-path", "", "path to the gitconfig file")
+	cmdutil.AddPrinterFlags(cmd)
+
+	return cmd
+}
+
+func (o *CreateSourceSecretOptions) CreateSourceSecret() error {
+	secret, err := o.MakeSourceSecret()
+	if err != nil {
+		return err
+	}
+
+	if _, err := o.SecretsInterface.Create(secret); err != nil {
+		return err
+	}
+
+	// fmt.Fprintf(o.GetOut(), "secret/%s\n", secret.Name)
+
+	return nil
+}
+
+type SourceConfigEntry struct {
+	// Username    string
+	Password    string
+	Token       string
+	Certificate string
+}
+
+const (
+	GitConfigKey = ".gitconfig"
+)
+
+func (o *CreateSourceSecretOptions) MakeSourceSecret() (*api.Secret, error) {
+
+	ca_content, err := ioutil.ReadFile(o.CertificatePath)
+	if err != nil {
+		return nil, err
+	}
+
+	gitAuth := SourceConfigEntry{
+		// Username:    o.Username,
+		Password:    o.Password,
+		Token:       o.Token,
+		Certificate: string(ca_content),
+	}
+
+	gitConfig := map[string]SourceConfigEntry{o.Username: gitAuth}
+
+	gitConfigContent, err := json.Marshal(gitConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	secret := &api.Secret{}
+	secret.Namespace = o.SecretNamespace
+	secret.Name = o.SecretName
+	secret.Type = api.SecretTypeOpaque
+	secret.Data = map[string][]byte{}
+	secret.Data[GitConfigKey] = gitConfigContent
+
+	return secret, nil
+}
+
+func (o *CreateSourceSecretOptions) Complete(f *cmdutil.Factory, args []string) error {
+	for i := range args {
+		fmt.Printf(args[i])
+	}
+
+	if len(args) != 1 {
+		return errors.New("must have exactly one argument: secret name")
+	}
+	o.SecretName = args[0]
+
+	client, err := f.Client()
+	if err != nil {
+		return err
+	}
+	o.SecretNamespace, _, err = f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	o.SecretsInterface = client.Secrets(o.SecretNamespace)
+
+	return nil
+}
+
+func (o CreateSourceSecretOptions) Validate() error {
+	if len(o.GitConfigPath) != 0 {
+		err := o.CheckGitConfig()
+		if err != nil {
+			return err
+		}
+	}
+
+	switch {
+	case len(o.SecretNamespace) == 0:
+		return errors.New("source secret namespace must be present")
+	case len(o.SecretName) == 0:
+		return errors.New("source secret name must be present")
+	case len(o.Username) == 0:
+		return errors.New("source secret username must be present")
+	case len(o.Password) == 0 && len(o.Token) == 0:
```

Comment:
Well in the .gitconfig you can have both, although token should have higher priority when cloning